src/dragonball/src/dbs_pci/src/msi.rs

```diff
@@ -0,0 +1,759 @@
+                }
+            }
+            4 => {
+                let value = LittleEndian::read_u32(data);
+                match offset {
+                    0x0 => {
+                        self.msg_ctl = (self.msg_ctl
+                            & !(MSI_CTL_ENABLE | MSI_CTL_MULTI_MSG_ENABLE))
+                            | ((value >> 16) as u16 & (MSI_CTL_ENABLE | MSI_CTL_MULTI_MSG_ENABLE));
+                    }
+                    MSI_MSG_ADDR_LO_OFFSET => {
+                        self.msg_addr_lo = value & MSI_MSG_ADDR_LO_MASK;
+                    }
+                    x if addr_hi_offset.is_some() && x == addr_hi_offset.unwrap() => {
+                        self.msg_addr_hi = value;
+                    }
+                    x if x == msg_data_offset => {
+                        self.msg_data = value as u16;
+                    }
+                    x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                        self.mask_bits = value;
+                    }
+                    _ => error!("invalid offset"),
+                }
+            }
+            _ => debug!("invalid data length"),
+        }
+
+        Ok(())
+    }
+
+    fn get_pending_state(&mut self, data: &mut [u8], offset: usize) {
+        for ptr in data.iter_mut() {
+            *ptr = 0;
+        }
+
+        if self.enabled() && self.per_vector_mask() {
+            if let Some(group) = self.group.as_ref() {
+                let start = offset * 8;
+                let end = std::cmp::min((offset + data.len()) * 8, group.len() as usize);
+
+                for idx in start..end {
+                    if self.vector_masked(idx as u16)
+                        && group.get_pending_state(idx as InterruptIndex)
+                    {
+                        data[idx / 8 - offset] |= 0x1u8 << (idx % 8);
+                    }
+                }
+            }
+        }
+    }
+
+    // Calculate message data offset depending on the address being 32 or 64 bits.
+    // Calculate upper address offset if the address is 64 bits.
+    // Calculate mask bits offset based on the address being 32 or 64 bits
+    // and based on the per vector masking being enabled or not.
+    fn get_offset_info(&self) -> (u64, Option<u64>, Option<u64>) {
+        if self.addr_64_bits() {
+            let mask_bits = if self.per_vector_mask() {
+                Some(0x10)
+            } else {
+                None
+            };
+            (0xc, Some(0x8), mask_bits)
+        } else {
+            let mask_bits = if self.per_vector_mask() {
+                Some(0xc)
+            } else {
+                None
+            };
+            (0x8, None, mask_bits)
+        }
+    }
+}
+
+impl PciCapability for MsiCap {
+    fn len(&self) -> usize {
+        self.size() as usize
+    }
+
+    fn set_next_cap(&mut self, next: u8) {
+        self.cap_id_next &= 0xff;
+        self.cap_id_next |= (next as u16) << 8;
+    }
```

Comment:
Add some comments for layout of cap? For example: `| CAP ID (high 8 bits) | OTHERS (low 8 bits) |`.

Reply:
comment is added for `MsiCap` structure.